### DISCUSSION TITLE
General testing of latest build

### DISCUSSION THREAD

**Discussion by gskjold** [UtilitechAS/amsreader-firmware#529] [General]
There have been some strange problems lately that have kept me busy in the code, so I have done a number of changes/improvements along the way. Would like to have some more people running this for a day or two before releasing.

If you install this firmware, please let me know your country, type of device and meter manufacturer.

- General optimizations
  - Moved all static strings to progmem to reduce use of heap memory
  - Split larger methods into multiple small methods to reduce use of stack memory
  - Use float instead of double for faster calculations where possible
- Improved firmware upgrade
  - Increased timeout to improve chance of upgrade via URL
  - Store expected firmware version and show warning in system page if upgrade was unsuccessful
- Dynamic RX buffer on HAN port
  - Keeping track on successful reads vs RX buffer overflow, increases buffer if necessary
- Improved debugging
  - Debug info if loop() method spends more than 2000ms on a single loop
  - Additional warning on several methods used in loop if they spend more than 1000ms
- Fixed filter on SSID to allow wider range of characters
- Fixed reboot loop after uploading config file on ESP32
- Improved handling of error codes from price API
- General changes along the way
 
See code changes here: https://github.com/UtilitechAS/amsreader-firmware/compare/v2.2.13...master

[esp32solo.zip](https://github.com/UtilitechAS/amsreader-firmware/files/11269588/esp32solo.zip)
[esp32.zip](https://github.com/UtilitechAS/amsreader-firmware/files/11269589/esp32.zip)
[esp8266.zip](https://github.com/UtilitechAS/amsreader-firmware/files/11269590/esp8266.zip)
[esp32c3.zip](https://github.com/UtilitechAS/amsreader-firmware/files/11269591/esp32c3.zip)
[esp32s2.zip](https://github.com/UtilitechAS/amsreader-firmware/files/11269592/esp32s2.zip)

**Comment by rogerthn2019** [comment]
- Fixed reboot loop after uploading config file on ESP32

Confirmed

**Comment by bjornsivertsen** [comment]
> Fixed filter on SSID to allow wider range of characters

Confirmed on Kamstrup in Denmark on a 8266 board. Everything else seems to be working fine!

**Comment by gskjold** [comment]
Thanks for testing everyone, release is on its way
